### ISSUE TITLE
Torch interface in Class?

### ISSUE THREAD

**Issue by ycchen1989** [XanaduAI/pennylane#351]
Hello everyone,
When I implemented the quantum circuits into Class, it seems that the torch interface not working properly. The same code works fine when it is not in a Class. Just wondering if anyone has the same problem? 
Thanks a lot!

`

	def _layer(self, W):
		""" Single layer of the variational classifier.

		Args:
			W (array[float]): 2-d array of variables for one layer
		"""
		qml.CNOT(wires=[0, 1])
		qml.CNOT(wires=[1, 2])
		qml.CNOT(wires=[2, 3])
		qml.CNOT(wires=[3, 4])
		qml.CNOT(wires=[4, 5])
		qml.CNOT(wires=[5, 6])
		qml.CNOT(wires=[6, 7])
		qml.CNOT(wires=[7, 8])
		qml.CNOT(wires=[8, 9])

		qml.Rot(W[0, 0], W[0, 1], W[0, 2], wires=0)
		qml.Rot(W[1, 0], W[1, 1], W[1, 2], wires=1)
		qml.Rot(W[2, 0], W[2, 1], W[2, 2], wires=2)
		qml.Rot(W[3, 0], W[3, 1], W[3, 2], wires=3)
		qml.Rot(W[4, 0], W[4, 1], W[4, 2], wires=4)
		qml.Rot(W[5, 0], W[5, 1], W[5, 2], wires=5)
		qml.Rot(W[6, 0], W[6, 1], W[6, 2], wires=6)
		qml.Rot(W[7, 0], W[7, 1], W[7, 2], wires=7)
		qml.Rot(W[8, 0], W[8, 1], W[8, 2], wires=8)
		qml.Rot(W[9, 0], W[9, 1], W[9, 2], wires=9)

	def circuit(self, angles = None):

		@qml.qnode(self.dev, interface='torch')
		def _circuit(angles = None):
			"""The circuit of the variational classifier."""
			qml.QubitStateVector(angles, wires=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

			weights = self.var_Q_circuit
			
			for W in weights:
				self._layer(W)

			return [qml.expval.PauliZ(0),qml.expval.PauliZ(1)]

		return _circuit(angles = angles)
`
Error message:
Rot: Real scalar parameter expected, got <class 'torch.Tensor'>.

**Comment by co9olguy** [comment]
@ycchen1989 Can you share the part of your code where you use a class? The code you posted only contains functions

**Comment by ycchen1989** [comment]
The complete Class is,

`

class VariationalQuantumClassifierBinary:
	def __init__(
			self,
			var_Q_circuit = None,
			var_Q_bias = None):

		self.var_Q_circuit = var_Q_circuit
		self.var_Q_bias = var_Q_bias
		self.dev = qml.device('default.qubit', wires=10)

	def set_params(self, var_Q_circuit, var_Q_bias):
		self.var_Q_circuit = var_Q_circuit
		self.var_Q_bias = var_Q_bias

	def _layer(self, W):
		""" Single layer of the variational classifier.

		Args:
			W (array[float]): 2-d array of variables for one layer

		"""

		qml.CNOT(wires=[0, 1])
		qml.CNOT(wires=[1, 2])
		qml.CNOT(wires=[2, 3])
		qml.CNOT(wires=[3, 4])
		qml.CNOT(wires=[4, 5])
		qml.CNOT(wires=[5, 6])
		qml.CNOT(wires=[6, 7])
		qml.CNOT(wires=[7, 8])
		qml.CNOT(wires=[8, 9])

		qml.Rot(W[0, 0], W[0, 1], W[0, 2], wires=0)
		qml.Rot(W[1, 0], W[1, 1], W[1, 2], wires=1)
		qml.Rot(W[2, 0], W[2, 1], W[2, 2], wires=2)
		qml.Rot(W[3, 0], W[3, 1], W[3, 2], wires=3)
		qml.Rot(W[4, 0], W[4, 1], W[4, 2], wires=4)
		qml.Rot(W[5, 0], W[5, 1], W[5, 2], wires=5)
		qml.Rot(W[6, 0], W[6, 1], W[6, 2], wires=6)
		qml.Rot(W[7, 0], W[7, 1], W[7, 2], wires=7)
		qml.Rot(W[8, 0], W[8, 1], W[8, 2], wires=8)
		qml.Rot(W[9, 0], W[9, 1], W[9, 2], wires=9)


	def circuit(self, angles):

		@qml.qnode(self.dev, interface='torch')
		def _circuit(angles):
			"""The circuit of the variational classifier."""
			qml.QubitStateVector(angles, wires=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

			weights = self.var_Q_circuit
			
			for W in weights:
				self._layer(W)

			return [qml.expval.PauliZ(0),qml.expval.PauliZ(1)]

		return _circuit(angles)

	def _forward(self, angles):
		"""The variational classifier."""
		# Output is torch Tensor

		# weights = self.var_Q_circuit
		bias = self.var_Q_bias 

		# Normalization the data vector before send into the state preparation
		# angles = angles / torch.sqrt(torch.sum(angles ** 2))

		angles = angles / torch.sqrt(torch.sum(angles ** 2))

		# print("ANGLES: ", angles)

		# print("ANGLES in VC:")
		# print(angles)

		# raw_output = torch.tensor(circuit(weights, angles=angles), dtype = torch.double)
		raw_output = self.circuit(angles)

		m = nn.Softmax(dim=0)

		normalized_output = torch.max(raw_output + bias, torch.tensor([1e-9,1e-9], device=device).type(dtype))

		output = m(normalized_output)
		
		return output

	def forward(self, angles):
		# Output torch tensor

		# angles = angles[0]
		# angles = torch.from_numpy(angles)
		# fw = np.array([self._forward(angles).detach().numpy()])
		# fw = self._forward(angles).detach()
		fw = self._forward(angles)
		# print("FW:",fw)
		return fw



`

**Comment by antalszava** [comment]
Hi @ycchen1989, thanks for submitting your issue!

In the version that you have posted in function ```_circuit(angles)``` the weights used for each layer are obtained by ```weights = self.var_Q_circuit```. Therefore ```weights``` is going to contain ```Tensor``` objects.  To use Pennylane's PyTorch interface and hence to be able to pass ```Tensor``` objects to circuits, you'll need ```weights``` in the signature of ```_circuit```, just like ```angles```. This way they will then be converted to ```Variable``` objects.

Could you please try with the following ```circuit``` defined in your class?

```python
    def circuit(self, angles):
        @qml.qnode(self.dev, interface='torch')
        def _circuit(weights, angles):
            """The circuit of the variational classifier."""
            qml.QubitStateVector(angles, wires=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

            for W in weights:
                self._layer(W)

            return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]

        return _circuit(self.var_Q_circuit, angles)
```
Another (minor) modification is the way the expectation value is obtained, the newer version of Pennylane supports the following syntax: ```qml.expval(qml.PauliZ(0))```.

**Comment by ycchen1989** [comment]
> Hi @ycchen1989, thanks for submitting your issue!
> 
> In the version that you have posted in function `_circuit(angles)` the weights used for each layer are obtained by `weights = self.var_Q_circuit`. Therefore `weights` is going to contain `Tensor` objects. To use Pennylane's PyTorch interface and hence to be able to pass `Tensor` objects to circuits, you'll need `weights` in the signature of `_circuit`, just like `angles`. This way they will then be converted to `Variable` objects.
> 
> Could you please try with the following `circuit` defined in your class?
> 
> ```python
>     def circuit(self, angles):
>         @qml.qnode(self.dev, interface='torch')
>         def _circuit(weights, angles):
>             """The circuit of the variational classifier."""
>             qml.QubitStateVector(angles, wires=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
> 
>             for W in weights:
>                 self._layer(W)
> 
>             return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]
> 
>         return _circuit(self.var_Q_circuit, angles)
> ```
> 
> Another (minor) modification is the way the expectation value is obtained, the newer version of Pennylane supports the following syntax: `qml.expval(qml.PauliZ(0))`.

It works! Thanks a lot!

**Comment by antalszava** [comment]
Great! Feel free to reach out, should you have further questions/concerns! :)
